Update ERC-7930: Move to Review

clowestab opened the pull request:
Updates to ERC-7930 to make it more clear and accessible based on community feedback and discussions within the Interop working group.

When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

0xrcinus commented on ERCS/erc-7930.md:
```diff
@@ -2,144 +2,127 @@
+## Comparisons with other standards
 
-### Comparisons with other standards
+### CAIP-10
```
```suggestion
#### CAIP-10
```

0xteddybear commented on ERCS/erc-7930.md:
```diff
@@ -2,144 +2,136 @@
 
-### Interoperable Address V1 binary format definition
-In binary format, addresses MUST have the following encoding:
+An _Interoperable Address_ as defined by this standard MUST have the following binary format:
```
Good fix on using the name whenever applicable

0xteddybear commented on ERCS/erc-7930.md:
```diff
@@ -2,144 +2,136 @@
-Address
-: Variable length field containing the binary format of the address component. For serialization details of different types of addresses see the [CAIP-350] profile for the chain type.
+**Address**
+: Variable length field containing the binary encoding of the address. To discern how to serialize an address for a specific chain type, implementors MUST reference the respective [CAIP-350] profile.
```
nit: I think it makes sense to normative-MUST the algorithm complying with the CAIP profile, and not the implementor referencing the spec

clowestab replied:
Ha. Valid point. Updated.

0xteddybear commented on ERCS/erc-7930.md:
```diff
@@ -2,144 +2,136 @@
+Clients MAY include the checksum when displaying an _Interoperable Name_ within their interface.
+Clients MAY accept _Interoperable Name_ inputs without a checksum.
+
+## Examples
```
iirc EIP-1 requires this is called test cases

0xteddybear reviewed (changes requested):
thanks for iterating on this!

Some of the edits do make the document easier to read, however some sections lost a great deal of normative power (e.g. the definition of the interoperable names) and I'd like for this to be as clear-cut of a spec as possible. The better-flowing prose is very well suited for an explainer-type document, though.

Also, some editorial concerns:
- the reordering breaks automated checks on the document structure (e.g. no extraneous sections, required sections, etc)
- the status should be moved to Review, and then we should ask an editor to move it to Last Call

clowestab commented on ERCS/erc-7930.md:
```diff
@@ -205,25 +210,38 @@ Interoperable Address:
-[ERC-7785]: ./eip-7785.md
 [CAIP-2]: https://github.com/ChainAgnostic/CAIPs/blob/2a7d42aebaffa42d1017c702974395ff5c1b3636/CAIPs/caip-2.md
 [CAIP-10]: https://github.com/ChainAgnostic/CAIPs/blob/2a7d42aebaffa42d1017c702974395ff5c1b3636/CAIPs/caip-10.md
 [CAIP-50]: https://github.com/ChainAgnostic/CAIPs/blob/2a7d42aebaffa42d1017c702974395ff5c1b3636/CAIPs/caip-50.md
```
👍 CAIP-350 is linked.

clowestab commented on ERCS/erc-7930.md:
```diff
@@ -205,25 +219,38 @@ Interoperable Address:
+## Rationale
 
-It is therefore advised for implementers requiring canonicity of addresses (e.g by using them as keys in smart contract mappings or other key-value stores), to thoroughly review the [CAIP-350] profile of a chain namespace for the possibility of a lack of canonicity of addresses (which should be noted in the profile's 'Extra Considerations' section) as well as collisions with other already-support ...
+- The `@` symbol is used as a separator as it is clear, easy for software to parse, and avoids confusion with the colon (`:`) symbol utilized in [CAIP-2] identifiers. 
```
Specifically as it pertains to canonicity as a security consideration? Happy to flesh out - is there an obvious example of a collision with another namespace?

clowestab commented:
> thanks for iterating on this!
> 
> Some of the edits do make the document easier to read, however some sections lost a great deal of normative power (e.g. the definition of the interoperable names) and I'd like for this to be as clear-cut of a spec as possible. The better-flowing prose is very well suited for an explainer-type document, though.
> 
> Also, some editorial concerns:
> 
> * the reordering breaks automated checks on the document structure (e.g. no extraneous sections, required sections, etc)
> * the status should be moved to Review, and then we should ask an editor to move it to Last Call

I've updated, and commented as appropriate.

I've moved it to 'Review' and fixed the validation checks.

I think my fundamental issue was that on first read I was unable to follow exactly what was going on. The separation of the _Interoperable Address_ and _Interoperable Name_ definitions in my opinion makes this significantly more accessible. 

I do agree that the spec should be as clear-cut as possible, so if I've inadvertently removed meaning I apologise, and am happy to iterate to make sure it is all in there.

This is my first foray into contributing to specs, and my perception is that a spec should be easy to understand independent of any external resources or explainers so as to incentivize people to actually read and contribute.

0xteddybear commented on ERCS/erc-7930.md:
```diff
@@ -2,144 +2,136 @@
+- Is compact for usage with cross-chain message passing and intent declaration.
+- Includes short, easily verififiable checksums to protect users.
+- Extends beyond EVM blockchains.
+- Is flexible and extensible
```
I believe this is a bit easier to follow on _what_ this standard provides, but not why/how. I wrote it like this to address actual questions from people in the ecosystem (with @frangio in particular being the guinea pig on how to sell a smart contract implementor on this). My impression is that every standard sells itself as flexible and extensible, and it's crucial to explain what it means in this context

0xteddybear commented on ERCS/erc-7930.md:
```diff
@@ -205,25 +219,40 @@ Interoperable Address:
+- MUST support defining an address, a chain, or both
+- MUST allow the client to discern the data necessary to calculate the checksum as outlined in this standard
+- MAY only be able to represent a subset of the CAIP namespaces
+- MAY expand upon the _Interoperable Name_ definition
```
'expand' is much less specific than 'assign extra syntactic restrictions', what do you mean precisely by 'expand'

clowestab replied:
'assign extra syntactic restrictions' feels constraining in that 7828 for example actually loosens restrictions - it allows you to use ENS names. My usage of 'expand' was I guess intentionally less specific as we don't know _how_ people may build upon this in the future.

I've gone for '_MAY modify the syntactic restrictions of the _Interoperable Name_ definition_' for now..

0xteddybear commented on ERCS/erc-7930.md:
```diff
@@ -205,25 +219,40 @@ Interoperable Address:
 
-## Copyright
-Copyright and related rights waived via [CC0](../LICENSE.md).
+- The `@` symbol is used as a separator as it is clear, easy for software to parse, and avoids confusion with the colon (`:`) symbol utilized in [CAIP-2] identifiers. 
```
software parsing is not particularly hard in the case of using another separator, but might be less obvious for humans

clowestab replied:
Updated. Still left in '_easy for software to parse_' because it's still true, even if other separators are just as easy.

0xteddybear commented on ERCS/erc-7930.md:
```diff
@@ -205,25 +219,40 @@ Interoperable Address:
+- No length restriction is placed on addresses, allowing for chains with longer address formats to be represented. This is flexible and future proof.
+- The address field includes `%` as a valid character to allow for URL-encoded address formats.
+- We chose to support zero-length addresses and chain IDs to make this standard flexible and to allow developers to use a single, uniform standard for many different jobs.
+- We chose **not** to utilize alternate encoding formats (`base58` or `base64` for example) to reduce the size of the _Interoperable Address_ and/or to abstract away namespace specific address serialization standards. We opted not to noting that readability and familiarity are important user experience considerations.
```
'we opted not to noting that readability...' not sure I follow this sentence

clowestab replied:
I was trying to explain our justification - that the space savings are not worth making it less familiar or readable for users. I have clarified further.

0xteddybear reviewed (changes requested):
kudos on complying with the editorial requirements, but there is still a significant loss of specificity

clowestab commented on ERCS/erc-7930.md:
```diff
@@ -205,25 +219,40 @@ Interoperable Address:
+This standard aims to be a foundational standard to allow for canonical representations of chain specific addresses. In the real world it can not be guaranteed that there will only be one valid representation of a chain specific address. For example, a particular chain namespace might define a serialization scheme that can't guarantee canonicity of addresses, or a given network might have two val ...
 
--->
+It is therefore advised that implementers requiring canonicity of addresses (if for example you are using them as keys in smart contract mappings), to thoroughly review the [CAIP-350] profile of a chain namespace for the possibility of a lack of canonicity. These considerations should be noted in the profile's 'Extra Considerations' section.
```
Not sure what happened there. Removed from 'Rationale' as more appropriate in 'Security Considerations'.

0xMonoAx commented on ERCS/erc-7930.md:
```diff
@@ -2,144 +2,142 @@
+
+A 4-byte checksum MUST be computed and included when sharing an _Interoperable Name_. 
+
+If a user-provided _Interoperable Name_ includes a checksum, clients MUST discern the underlying _Interoperable Address_ recompute the checksum and compare it to the provided value. On mismatch, clients MUST alert the user.
```
What do you think about adding this section? I think it would improve clarity regarding the relationship with CAIP-350 (if I add this, I would remove the clarification from the Abstract)

_**Parsing and Serialization:**_

_When parsing an Interoperable Name to generate its binary Interoperable Address, clients MUST follow the normalization and serialization rules defined in the relevant CAIP-350 profile for the given `<chain>` and `<address>.` This ensures that different valid text representations (e.g., case variations in an address) resolve to a single, canonical binary form, which is essential for consistent checksum calculation and data integrity._

0xteddybear commented on ERCS/erc-7930.md:
```diff
@@ -2,144 +2,146 @@
 title: Interoperable Addresses
 description: An extensible binary format to refer to an address specific to one chain.
-author: Teddy (@0xteddybear), Joxes (@0xJoxess), Nick Johnson (@Arachnid), Francisco Giordano (@frangio), Skeletor Spaceman (@skeletor-spaceman), Racu (@0xRacoon), TiTi (@0xtiti), Gori (@0xGorilla), Ardy (@0xArdy), Onizuka (@onizuka-wl), Sam Kaufman (@SampkaML), Marco Stronati (@paracetamolo), Yuliya Alexiev (@yuliyaalexiev), Jeff Lau (@jefflau), Sam Wilson (@samwilsn), Vitalik Buterin (@vbuterin ...
+author: Teddy (@0xteddybear), Joxes (@0xJoxess), Nick Johnson (@Arachnid), Francisco Giordano (@frangio), Skeletor Spaceman (@skeletor-spaceman), Racu (@0xRacoon), TiTi (@0xtiti), Gori (@0xGorilla), Ardy (@0xArdy), Onizuka (@onizuka-wl), Sam Kaufman (@SampkaML), Marco Stronati (@paracetamolo), Yuliya Alexiev (@yuliyaalexiev), Jeff Lau (@jefflau), Sam Wilson (@samwilsn), Vitalik Buterin (@vbuterin ...
```
mono's username is missing the `@`

0xteddybear commented on ERCS/erc-7930.md:
```diff
@@ -2,144 +2,146 @@
-: A binary payload which unambiguously identifies a target address and allows conversion to a human-readable name.
+### Terminology
+**Target Address**
+: The target address being represented. Serialized subject to the rules outlined in the [CAIP-350] profile for the namespace of the chain being targeted.
```
so far we have used 'target address' to mean an address already scoped to a given chain, aka the object of this proposal, and we haven't used the 'target chain' concept. Also, this definition is recursive :grin:

0xteddybear commented on ERCS/erc-7930.md:
```diff
@@ -2,144 +2,146 @@
-    - It makes the Interoperable Address <-> raw address relationship evident to the user, which will make the cases where the former are not (yet) supported by a dapp or hardware wallet easier to reason with, as users will evidently see the address is the same, but it's the chain&checksum that is being trimmed.
-    - Its extra human-readability will make it more useful as a a stopgap solution until a standard focused on chain and address names is finalized.
-- We chose to support zero-length addresses and chainids to be able to use just one standard to represent both potentially foreign _target addresses_ (e.g. the recipient of a cross-chain message) and also plain addresses or chainids (e.g. to specify the origin-chain hook address of a cross-chain message), for greater uniformity across implementations.
+The 4-byte `<checksum>` is the first 4 bytes of the hexadecimal ([RFC 4648]) string representation of a keccak256 hash. This hash is computed over the concatenation of the _Interoperable Address_'s binary fields: `ChainType`, `ChainReferenceLength`, `ChainReference`, `AddressLength`, and `Address`. The `Version` field MUST NOT be included in the hashed data.
```
nit: 'the 4byte checksum is the first 4 bytes...' feels like could be improved writing-wise, but can't think of an actually better alternative. It does not compromise clarity though, so not a blocker

0xteddybear commented on ERCS/erc-7930.md:
```diff
@@ -205,25 +229,38 @@ Interoperable Address:
--->
+- The `@` symbol is used as a separator as it provides visual clarity to humans, is easy for software to parse, and avoids confusion with the colon (`:`) symbol utilized in [CAIP-2] identifiers. 
+- No length restriction is placed on addresses, allowing for chains with longer address formats to be represented.
+- The address field includes `%` as a valid character to allow for URL-encoded address formats.
```
the previous expression was clearer than this one: 

>  Similarly, the address field includes `%` as a valid character to allow for url-encoding of any other characters.

it clearly states that the purpose of using `%` is to url-encode characters not natively supported by the standard. On the other hand, it's not clear what 'URL-encoded' address formats' means, precisely. Are there address formats out there which are in some way URLs?

0xteddybear commented on ERCS/erc-7930.md:
```diff
@@ -205,25 +229,38 @@ Interoperable Address:
+## Rationale
 
--->
+- The `@` symbol is used as a separator as it provides visual clarity to humans, is easy for software to parse, and avoids confusion with the colon (`:`) symbol utilized in [CAIP-2] identifiers. 
```
the previous wording is a more comprehensive explaination of _why_ we chose to use the `@` character. In particular:
- it makes it trivial to distinguish CAIP-10 ids from Interoperable Names
- it also makes sure we can fit into this format anything that can fit in a CAIP-10
